cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,315 @@
+     "-'    "--                                                            
+                                                                           
+----------------------------------------------------------------------------
+
+  ******************************************************************************
+  * @file    dma.c
+  * @author  MCD Application Team, Wi6labs
+  * @brief   dma service.
+  ******************************************************************************
+
+Copyright (c) <2017>, <Wi6labs>
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the wi6labs nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL WI6LABS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+describe file content here
+-----------------------------------------------------------------------------*/
+#include "stm32_def.h"
+#include "dma.h"
+#include "analog.h"
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+extern ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
+
+static DMA_HandleTypeDef DmaHandleTab[ADC_NUM] = {NULL};
+static bool dmaTab[DMA_NUMBER][CHANNEL_NUMBER];
+
+/**
+  * @brief  This function will init the dma and channel corresponding 
+  *         to the ADC peripheral
+  * @param  pin : the ADC peripheral to use
+  * @retval none
+  */
+void dma_init_adc(ADC_HandleTypeDef *hadc)
+{
+  //DMA_HandleTypeDef  DmaHandle;
+  void*     channel;
+  IRQn_Type irqHandler;
+  int8_t    channelNumber;
+  int8_t    dmaNumber;
+  uint8_t   indice = 0;
+  
+#ifdef __HAL_RCC_DMA1_CLK_ENABLE
+  __HAL_RCC_DMA1_CLK_ENABLE();
+#endif
+
+#ifdef __HAL_RCC_DMA2_CLK_ENABLE
+  __HAL_RCC_DMA2_CLK_ENABLE();
+#endif
+
+  channel = pinmap_find_dmaChannel(hadc->Instance, DmaMap_DMA);
```

Comment:
Instead of DmaMap_DMA a generic table can be used, where any kind of devices Serial, ADC could be listed with corresponding ADC instance and channel. ... 
Then the table could be parsed with a for loop instead of checks like (if (hadc->Instance == ADC3))

Reply:
pinmap_find_dmaChannel is similar to other functions like pin_pinName or pinmap_find_function. Why should it be different? DmaMap_DMA contains all informations needed for DMA like PinMap_ADC contains all informations needed for ADC.

Reply:
One important thing: PeripheralsPins.c contains arrays for pin mapping.
What a pin is able to do for a dedicated peripheral and how.
So adding this kind of array not linked to a pin is not correct.
+const DmaMap DmaMap_DMA[] = {
+    {ADC1, 1, DMA1_Channel1, DMA1_Channel1_IRQn        , 1},
+    {ADC1, 1, DMA1_Channel2, DMA1_Ch2_3_DMA2_Ch1_2_IRQn, 2},
+    {ADC1, 2, DMA2_Channel5, DMA1_Ch4_7_DMA2_Ch3_5_IRQn, 5},
+    {NP,   0, NP,            NP,                         0}
+};

Reply:
A DMA is link to a peripheral not a pin... i agree we that. Can we keep this table an put it in another file? Oterwise, what do you propose?

Reply:
Currently, I do not reviewed this PR so I have no proposal. Just a note on PeripheralPins.c file usage.

Reply:
I agree with proposal to keep the table that is needed but move it to dma.c as this is a table about dma channels and streams, not pins  - is it ok ?

Reply:
The table if different from one board to another. If we move it to dma.c, we will frame it with #ifdef. dma.c will become a huge file. Is it ok for you?